### PR TITLE
Update to support prettier 3.0

### LINKED ISSUE #2153: `@typespec/prettier-plugin-typespec` errors with Prettier 3.0.0
Prettier 3.0.0 was released yesterday, and throws errors when attempting to format any TypeSpec files.

Looking at a minimised snippet from one of the samples:

```
// main.tsp
import "@typespec/versioning";
import "@typespec/rest";
import "./library.tsp";

using TypeSpec.Versioning;
using TypeSpec.Rest;

@service({
  title: "Pet Store Service",
})
@versioned(Versions)
namespace VersionedApi;
using TypeSpec.Http;

enum Versions {
  @useDependency(Library.Versions.`1.0`)
  v1,

  @useDependency(Library.Versions.`1.1`)
  v2,
}
```

And a simple Prettier config:
```json
{
  "plugins": [
    "@typespec/prettier-plugin-typespec"
  ],
  "overrides": [
    {
      "files": "*.tsp",
      "options": {
        "parser": "typespec"
      }
    }
  ]
}
```

The plugin works with 2.8.7:
```
$ prettier --version
2.8.7

$ prettier --write main.tsp 
main.tsp 9ms
```

But throws errors with 3.0.0:
```
$ npx prettier --version
3.0.0

$ npx prettier --write main.tsp 
main.tsp
[error] main.tsp: InvalidDocError: Unexpected doc.type 'concat'.
[error] Expected it to be 'cursor', 'indent', 'align', 'trim', 'group', 'fill', 'if-break', 'indent-if-break', 'line-suffix', 'line-suffix-boundary', 'line', 'label', or 'break-parent'.
[error]     at traverseDoc (file:///data/src/2.CCF/node_modules/prettier/index.mjs:15105:13)
[error]     at propagateBreaks (file:///data/src/2.CCF/node_modules/prettier/index.mjs:15437:3)
[error]     at printDocToString (file:///data/src/2.CCF/node_modules/prettier/index.mjs:15842:3)
[error]     at coreFormat (file:///data/src/2.CCF/node_modules/prettier/index.mjs:18409:18)
[error]     at async formatWithCursor (file:///data/src/2.CCF/node_modules/prettier/index.mjs:18603:14)
[error]     at async formatFiles (file:///data/src/2.CCF/node_modules/prettier/internal/cli.mjs:6778:18)
[error]     at async main (file:///data/src/2.CCF/node_modules/prettier/internal/cli.mjs:6970:5)
[error]     at async Module.run (file:///data/src/2.CCF/node_modules/prettier/internal/cli.mjs:6924:5)
```

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #2153

This include breaking change if using the typespec formatter programatically. Some breaking changes to prettier carry over https://prettier.io/blog/2023/07/05/3.0.0.html

### CODE REVIEW

File: packages/compiler/test/emitter-framework/typescript-emitter.ts [hunk 1]
```diff
@@ -277,9 +277,9 @@ export class TypeScriptInterfaceEmitter extends CodeTypeEmitter {
       emittedSourceFile.contents += decl.value + "\n";
     }
 
-    emittedSourceFile.contents = prettier.format(emittedSourceFile.contents, {
-      parser: "typescript",
-    });
+    // emittedSourceFile.contents = prettier.format(emittedSourceFile.contents, {
```
Comment (timotheeguerin) [inline packages/compiler/test/emitter-framework/typescript-emitter.ts:280]: This is an issue as the emitter framework doesn't allow async here

File: packages/compiler/src/formatter/print/printer.ts [hunk 2]
```diff
@@ -65,17 +68,19 @@ import {
...
 import { commentHandler } from "./comment-handler.js";
 import { needsParens } from "./needs-parens.js";
+import { AstPath } from "./prettier-ast-path.js";
 import { DecorableNode, PrettierChildPrint, TypeSpecPrettierOptions } from "./types.js";
 
 const { align, breakParent, group, hardline, ifBreak, indent, join, line, softline } =
   prettier.doc.builders;
 
-const { isNextLineEmpty } = prettier.util;
+const { isNextLineEmpty } = prettier.util as any;
```
Comment (timotheeguerin) [inline packages/compiler/src/formatter/print/printer.ts:83]: Todo check that

File: packages/json-schema/src/json-schema-emitter.ts [hunk 3]
```diff
@@ -532,7 +532,7 @@ export class JsonSchemaEmitter extends TypeEmitter<Record<string, any>, JSONSche
     const toEmit: EmittedSourceFile[] = [];
 
     for (const sf of sourceFiles) {
-      const emittedSf = this.emitter.emitSourceFile(sf);
+      const emittedSf = await this.emitter.emitSourceFile(sf);
```
Comment (timotheeguerin) [inline packages/json-schema/src/json-schema-emitter.ts:535]: This is unfortunate breaking change of having to make `sourceFile()` async

File: .prettierrc.json
Comment (timotheeguerin) [inline .prettierrc.json:1]: TODO: vscode prettier integration is broken, investigate

Comment (timotheeguerin) [reply .prettierrc.json:1]: resolved by using specifying the `index.cjs` entrypoint in the prettierPath config

File: .prettierrc.json [hunk 4]
```diff
@@ -1,15 +1,15 @@
 {
   "arrowParens": "always",
+  "trailingComma": "es5",
```
Comment (timotheeguerin) [inline .prettierrc.json:3]: Filed #2159 to change to the new default of `all`

File: packages/json-schema/src/json-schema-emitter.ts (hunk 3 above)
Comment (timotheeguerin) [reply packages/json-schema/src/json-schema-emitter.ts:535]: moved that breaking change into its own PR https://github.com/microsoft/typespec/pull/2183

**Review by daviwil — APPROVED** [review]
Jeez!  Do we actually get any benefits from this upgrade?  Seems like they made you jump through hoops to integrate it.

**Comment by timotheeguerin** [comment]
Yeah there isn’t much benefits from consuming prettier 3.0. They added a few more targeted export so you don’t import the file system part in a plug-in but still missing the utils so do have to import from standalone. 

I’m off for a week so will revisit after to make sure they haven’t pushed a new update with some fixes which would simplify this.

But this lets people that use prettier 3.0 to be able to use the plugin.

**Comment by markcowl** [comment]
Change to discuss: formatTypeSpec helper is now async

**Comment by timotheeguerin** [comment]
> Change to discuss: formatTypeSpec helper is now async

Didn't we already discuss that 2 weeks ago?

**Pull request merged**